src/manifolds/VectorBundle.jl

```diff
@@ -804,6 +817,123 @@ function Random.rand!(rng::AbstractRNG, M::TangentSpaceAtPoint, X; vector_at=not
     return X
 end
 
+function _retract(M::VectorBundle, p, X, ::VectorBundleProductRetraction)
+    return retract_product(M, p, X)
+end
+
+function _retract!(M::VectorBundle, q, p, X, ::VectorBundleProductRetraction)
+    return retract_product!(M, q, p, X)
+end
+
+"""
+    retract_product(M::VectorBundle, p, q)
+
+Compute the allocating variant of the [`VectorBundleProductRetraction`](@ref),
+which by default allocates and calls `retract_product!`.
+"""
+function retract_product(M::VectorBundle, p, X)
+    q = allocate_result(M, retract, p, X)
+    return retract_product!(M, q, p, X)
+end
+
+function retract_product!(B::VectorBundle, q, p, X)
+    xp, Xp = submanifold_components(B.manifold, p)
+    xq, Xq = submanifold_components(B.manifold, q)
+    VXM, VXF = submanifold_components(B.manifold, X)
+    # this temporary avoids overwriting `p` when `q` and `p` occupy the same memory
+    xqt = exp(B.manifold, xp, VXM)
+    vector_transport_direction!(
+        B.manifold,
+        Xq,
+        xp,
+        Xp + VXF,
+        VXM,
+        B.vector_transport.method_point,
+    )
+    copyto!(B.manifold, xq, xqt)
+    return q
+end
+
+@doc raw"""
+    struct SasakiRetraction <: AbstractRetractionMethod end
+
+Exponential map on [`TangentBundle`](@ref) computed via Euler integration as described
+in [^Muralidharan2012]. The system of equations for $\gamma : ℝ \to T\mathcal M$ such that
+$\gamma(1) = \exp_{p,X}(X_M, X_F)$ and $\gamma(0)=(p, X)$ reads
+
+```math
+\dot{\gamma}(t) = (\dot{p}(t), \dot{X}(t)) = (R(X(t), \dot{X}(t))\dot{p}(t), 0)
+```
+
+where $R$ is the Riemann curvature tensor (see [`riemann_tensor`](@ref)).
+
+# Constructor
+
+    SasakiRetraction(L::Int)
+
+In this constructor `L` is the number of integration steps.
+
+[^Muralidharan2012]:
+    > P. Muralidharan and P. T. Fletcher, “Sasaki Metrics for Analysis of Longitudinal Data
+    > on Manifolds,” Proc IEEE Comput Soc Conf Comput Vis Pattern Recognit, vol. 2012,
+    > pp. 1027–1034, Jun. 2012, doi: 10.1109/CVPR.2012.6247780.
+"""
+struct SasakiRetraction <: AbstractRetractionMethod
```

Comment:
Could we put types/structs at the beginning of the file?

Reply:
Sure, I'll move it.